### DISCUSSION TITLE
Passing mocks to extension previously marked as @Shared

### DISCUSSION THREAD

**Discussion by reftel** [xvik/spock-junit5#84] [General]
Hi,

we´re using Spock together with [Dropwizard´s Junit 5 support](https://www.dropwizard.io/en/stable/manual/testing.html#junit-5), and we´d like to use spock-junit5 to switch away from manually calling `.before()` and `.after()` on the extensions to using spock-junit5. The extensions need objects from the application that in turn need mocks to talk to. We therefore have methods for building up a properly configured extension (see example below).
For tests where the extensions are per test method, this works perfectly. For tests where we create the extension once, and re-use it between methods, this is a lot trickier. Since spock-junit5 requires the field to be static, the methods for building the extensions now need to be static - and Spock´s mocking API does not seem to work from static methods.
For now, the only solution I´ve found is to use some other mocking library for the mocks that need to be passed to the extensions, but that´s so clunky it´s not really worth it. Are there any other ways this could be done?

Simplified example of setup of shared extension:
```groovy
@Shared
ResourceExtension resourceExtension = buildExtension()
def buildExtension() {
    MyService serviceMock = Mock() {
      1 * doTheThing() >> 42
    }
    MyResource resource = new MyResource(serviceMock)
    return ResourceExtension.builder()
        .addResource(resource)
        .build()
}
```

**Comment by xvik** [comment]
Spock-junit5 mimics junit 5 extensions behaviour to be sure that existing extensions would work the same way as in raw junit.

**Before** junit 5.11 it was required to declare `Before(After)All` (and test instance) extensions in static fields - extensions from non-static fields were registered after class instance initialization (too late for *All callbacks). 
But **in** 5.11 junit changed this behaviour and now it's not required to declare extensions in static fields. 

I just have time to find out this difference, but did not update spock-junit5 itself (that's why dependabot pull request with 5.11 was not accepted - some tests fail to validate old junit behaviour). Also, need to check how it would be compatible with spock lifecycle (could be surprises).

So probobly next spock-junit5 version would address your needs (can't be sure now). But can't promise when it would be released (as soon as I have time).

Probably, even better way would be to [write your own spock extensions](https://spockframework.org/spock/docs/2.3/extensions.html#_writing_custom_extensions): a simple wrapper around dropwizard extensions. Dropwizard extensions are generic and so not depend on junit 5 (their junit 5 extension actually just manually looks for their own extensions by interface and calls before/after methods in appropriate lifecycle phases).

**Comment by reftel** [comment]
Thank you for the answer! This actually also explains some (to me) puzzling behaviour I´ve seen when using class rules in Spock tests =)
We´ll consider the options. Now we have at least one way forward that does not involve using Mockito from a Spock test, very happy about that =)
